File: android/src/main/java/com/mcodex/nitroarchive/domain/DuplicatePathPolicy.kt
package com.mcodex.nitroarchive.domain

class DuplicatePathPolicy(private val policy: DuplicateEntryPolicy = DuplicateEntryPolicy.ERROR) {
    private val seenPaths = mutableSetOf<String>()

    fun check(path: String): Boolean {
        if (path in seenPaths) {
            return when (policy) {
                DuplicateEntryPolicy.ERROR -> throw ArchiveDomainError.DuplicateEntry("Duplicate output path: $path")
                DuplicateEntryPolicy.FIRST -> false
                DuplicateEntryPolicy.LAST -> true
            }
        }
        seenPaths.add(path)
        return true
    }
}

File: ios/Domain/DuplicatePathPolicy.swift
import Foundation

struct DuplicatePathPolicy {
  private var seenPaths: Set<String> = []
  private let policy: DuplicateEntryPolicy

  init(policy: DuplicateEntryPolicy = .error) {
    self.policy = policy
  }

  mutating func check(_ path: String) throws -> Bool {
    if seenPaths.contains(path) {
      switch policy {
      case .error:
        throw ArchiveDomainError.duplicateEntry("Duplicate output path: \(path)")
      case .first:
        return false
      case .last:
        return true
      }
    }
    seenPaths.insert(path)
    return true
  }
}
